datagateway_api/common/config.py

```diff
@@ -109,25 +88,85 @@ def set_backend_type(self, backend_type):
         type must be fetched. This must be done using this module (rather than directly
         importing and checking the Flask app's config) to avoid circular import issues.
         """
-        self._config["backend"] = backend_type
+        self.backend = backend_type
 
-    def get_icat_properties(self):
+    class Config:
         """
-        ICAT properties can be retrieved using Python ICAT's client object, however this
-        requires the client object to be authenticated which may not always be the case
-        when requesting these properties, hence a HTTP request is sent as an alternative
+        The behaviour of the BaseModel class can be controlled via this class.
         """
-        properties_url = (
-            f"{config.get_config_value(APIConfigOptions.ICAT_URL)}/icat/properties"
-        )
-        r = requests.request(
-            "GET",
-            properties_url,
-            verify=config.get_config_value(APIConfigOptions.ICAT_CHECK_CERT),
-        )
-        icat_properties = r.json()
 
-        return icat_properties
+        # Enables assignment validation on the BaseModel fields. Useful for when the
+        # backend type is changed using the set_backend_type function.
+        validate_assignment = True
+
+
+class SearchAPI(BaseModel):
+    """
+    Configuration model class that implements pydantic's BaseModel class to allow for
+    validation of the SearchAPI config data using Python type annotations.
+    """
+
+    client_pool_init_size: StrictInt
+    client_pool_max_size: StrictInt
+    extension: StrictStr
+    icat_check_cert: StrictBool
+    icat_url: StrictStr
+
+
+class TestUserCredentials(BaseModel):
+    username: StrictStr
+    password: StrictStr
+
+
+class APIConfig(BaseModel):
+    """
+    Configuration model class that implements pydantic's BaseModel class to allow for
+    validation of the API config data using Python type annotations. It ensures that
+    all required config options exist before getting too far into the setup of the API.
+
+    If a mandatory config option is missing or misspelled, or has a wrong value type,
+    Pydantic raises a validation error with a breakdown of what was wrong and the
+    application is exited.
+
+    Config options used for testing are not checked here as they should only be used
+    during tests, not in the typical running of the API.
+
+    Some options used when running the API (host, debug_mode etc.) aren't mandatory
+    when running the API in production (these options aren't used in the `wsgi.py`
+    entrypoint). As a result, they're not present in `config_keys`. However, they
+    are required when using `main.py` as an entrypoint. In any case of these
+    specific missing config options when using that entrypoint, they are checked at
+    API startup so any missing options will be caught quickly.
+    """
+
+    datagateway_api: Optional[DatagatewayAPI]
```

Comment:
Obviously this is marked by pydantic as optional (as it should be), but when I remove the `datagateway_api` object from `config.json`, the API errors on startup:

```bash
Traceback (most recent call last):
  File "/root/.pyenv/versions/3.6.8/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/root/.pyenv/versions/3.6.8/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/root/seg_dev/datagateway-api/datagateway_api/src/main.py", line 7, in <module>
    from datagateway_api.src.api_start_utils import (
  File "/root/seg_dev/datagateway-api/datagateway_api/src/api_start_utils.py", line 11, in <module>
    from datagateway_api.common.datagateway_api.backends import create_backend
  File "/root/seg_dev/datagateway-api/datagateway_api/common/datagateway_api/backends.py", line 4, in <module>
    from datagateway_api.common.datagateway_api.icat.backend import PythonICATBackend
  File "/root/seg_dev/datagateway-api/datagateway_api/common/datagateway_api/icat/backend.py", line 7, in <module>
    from datagateway_api.common.datagateway_api.icat.helpers import (
  File "/root/seg_dev/datagateway-api/datagateway_api/common/datagateway_api/icat/helpers.py", line 79, in <module>
    @cached(cache=ExtendedLRUCache())
  File "/root/seg_dev/datagateway-api/datagateway_api/common/datagateway_api/icat/lru_cache.py", line 22, in __init__
    super().__init__(maxsize=config.datagateway_api.client_cache_size)
AttributeError: 'NoneType' object has no attribute 'client_cache_size'
```

This is because of imports at the top of files, with `@cached(cache=ExtendedLRUCache())` making a call that looks at some DataGateway API specific config. Could you investigate if there's a way of refactoring the imports/code to stop this from happening please? The current behaviour means that a user cannot have config for just the search API. 

I wonder if this could be fixed by moving `get_cached_client()` into a class, maybe making that function into a static method? You might be able to think of a different/better solution so happy to see you fix it a different way.

Reply:
Decorators run during import, before the function is even called. So by having `@cached(cache=ExtendedLRUCache())`, it means that an `ExtendedLRUCache` object will always be instantiated at the beginning even if that is inside a class. What I did instead is I set the `create_backend` function to be imported in `api_start_utils.py` only if the config contains a `datagateway_api` object.